debug_toolbar/store.py

```diff
@@ -140,5 +141,91 @@ def panels(cls, request_id: str) -> Any:
             yield panel, deserialize(data)
 
 
+class DatabaseStore(BaseStore):
+    @classmethod
+    def _cleanup_old_entries(cls):
+        """
+        Enforce the cache size limit - keeping only the most recently used entries
+        up to RESULTS_CACHE_SIZE.
+        """
+        # Get the cache size limit from settings
+        cache_size = dt_settings.get_config()["RESULTS_CACHE_SIZE"]
+
+        # Determine which entries to keep (the most recent ones up to cache_size)
+        keep_ids = list(
+            HistoryEntry.objects.all()[:cache_size].values_list("request_id", flat=True)
+        )
+
+        # Delete all entries not in the keep list
+        if keep_ids:
+            HistoryEntry.objects.exclude(request_id__in=keep_ids).delete()
+
+    @classmethod
+    def request_ids(cls):
+        """Return all stored request ids within the cache size limit"""
+        cache_size = dt_settings.get_config()["RESULTS_CACHE_SIZE"]
+        return list(
+            HistoryEntry.objects.all()[:cache_size].values_list("request_id", flat=True)
+        )
+
+    @classmethod
+    def exists(cls, request_id: str) -> bool:
+        """Check if the given request_id exists in the store"""
+        return HistoryEntry.objects.filter(request_id=request_id).exists()
+
+    @classmethod
+    def set(cls, request_id: str):
+        """Set a request_id in the store and clean up old entries"""
+        # Create the entry if it doesn't exist (ignore otherwise)
+        HistoryEntry.objects.get_or_create(request_id=request_id)
+
+        # Enforce the cache size limit to clean up old entries
+        cls._cleanup_old_entries()
```

Comment:
```suggestion
        with transaction.atomic():
            # Create the entry if it doesn't exist (ignore otherwise)
            _, created = HistoryEntry.objects.get_or_create(request_id=request_id)
            if created:
                # Enforce the cache size limit to clean up old entries
                cls._cleanup_old_entries()
```
This does two things. It wraps the operation in a transaction so that we only drop create and drop old entries if everything works as expected. That may be overkill though.

The other part is that we only cleanup old entries if we're adding an entry. It reduces two queries.

Reply:
The transaction.atomic() wrapper seems overkill to me but I see the point about running cleanup only when new records are created, which we could do in the set method:


```python
    @classmethod
    def set(cls, request_id: str):
        """Set a request_id in the store and clean up old entries"""
        # Create the entry if it doesn't exist (ignore otherwise)
        _, created = HistoryEntry.objects.get_or_create(request_id=request_id)

        # Only enforce cache size limit when new entries are created
        if created:
            cls._cleanup_old_entries()
```

What do you think?

Reply:
Hmmm... this comment is showing under save_panel(), but looking at your next comment, I wonder if you were commenting on the set() method.